### ISSUE TITLE
Partial select 

### ISSUE THREAD

**Issue by saurabhabh** [mafintosh/hyperdb#156]
I have an use case where I plan to store the users in the db with key `users/{id}`. So the keys will be `users/1`
`users/2` 
and so on. 
Each key will have details for the users like name, address, phone number etc. 
Now I can retrieve all the users by using the _list api_  as follows
`db.list('/users/', function (err, nodes) { })`. 

Now Is there a way where I can select/get/list only 10 users from a list of 100 ?

**Comment by m-onz** [comment]
If you use db.list('/users/', function (err, nodes) you receive  callback with all matching nodes. It would be easy to loop through them and filter out what your not interested in using a condition. This could be done with loops or using map / forEach.

You can also try using a stream... 
```
var stream = db.createReadStream('/users')
stream.pipe(take(100)).pipe(process.stdout)
```

Where take(100) is a transform stream.

**Comment by saurabhabh** [comment]
Hi thanks for the reply. I am working on Angular - electron a desktop application. Here is what I implemented. I wanted the users in reverse order (last added , last modified) so I used the reverse option on the Stream. 
I dont seem to get the correct users. Can you shed some light on how the reverse works? does it work on time stamp ? I am looking to fetch last 5 users that are added to db. 

Below is the code snippet that I am using. 
 

`var stream = self.db.createReadStream(`/users/`, {reverse: true})`


            let  i = 0;
            let array:Array<User> = []
            stream.on('data', (chunk) => {
                 if( i >= 5){
                     stream.destroy();
                 }
                i = i + 1; 
                if(chunk && chunk.length > 0){
                    console.log(`Received ${chunk.length} bytes of data. ${chunk[0].value}`);
                    array.push(message);
                }
            });
            stream.on('end', function() {
                console.log("end");
                // sort the Users array by time.
                array =  _.sortBy(array, o => o.msgtime);
                cb(array);
            });
            stream.on('close', function() {
                console.log("close");
                // sort the Users array by time.
                array =  _.sortBy(array, o => o.msgtime);
                cb(array);
            });

**Comment by m-onz** [comment]
I normally add a timestamp and use map and sort to get what I want.

I've just verified what you mean, i'm not totally sure why its not sorting correctly even thought the sequence numbers look good.

```
var hyperdb = require('hyperdb')

var db = hyperdb('./test.db', { valueEncoding: 'json' })

var i = 0

var interval = setInterval(function () {
        db.put('/users/'+Date.now(), { incro: i }, function () {
                i++
        })
}, 1000)

setTimeout(function () {
        clearInterval(interval);
        var s = db.createReadStream('/users')
        s.on('data', console.log)
}, 9000)

```
returns..
```
[ Node(key=users/1548158437739, value={ incro: 5 }, seq=6, feed=0)) ]
[ Node(key=users/1548158435737, value={ incro: 3 }, seq=4, feed=0)) ]
[ Node(key=users/1548158434735, value={ incro: 2 }, seq=3, feed=0)) ]
[ Node(key=users/1548158433733, value={ incro: 1 }, seq=2, feed=0)) ]
[ Node(key=users/1548158432729, value={ incro: 0 }, seq=1, feed=0)) ]
[ Node(key=users/1548158436738, value={ incro: 4 }, seq=5, feed=0)) ]
[ Node(key=users/1548158438741, value={ incro: 6 }, seq=7, feed=0)) ]
[ Node(key=users/1548158439743, value={ incro: 7 }, seq=8, feed=0)) ]

```

**Comment by saurabhabh** [comment]
From the documentation it seems that `createHistoryStream` returns all the nodes historically. It doesnt seem to work with prefix. I found `createKeyHistoryStream` but again this does not support the reverse option. 
My whole purpose of getting in this is that I dont want to read the entire data at one go. And also want it sorted by timestamp desc. 
Suppose a use case of chats where the messages are more than 1000 between 2 users. Now I want to load only the last 100 added chat messages. How can I do it efficiently. ? 

- One way is to use the .list option iterate over them and then sort them via lodash.  

Is there some efficient option where I get a sorted list (desc by timestamp) and only the last 100 messages ? So that it decreases the read load on the db. ?

**Comment by m-onz** [comment]
It returns all the historical nodes... which is what you want. Its a readable stream so you can reverse it and take the first 100 items without downloading the entire dataset.

See my link to the streams handboook (specifically through streams)... you just need to push null onto the stream within a transform to end the stream gracefully (no need for stream.destroy())

**Comment by saurabhabh** [comment]
I store `/users/1` and also `message/234` the createHistoryStream not only return the users node but also the message node.
Is my understanding correct? I need only the users not the messages. That is where I need the 'prefix' I think.

**Comment by m-onz** [comment]
yes, ultimately it would be good to have the read stream from createReadStream('/users') sorted via time and then you can avoid having to do your own prefixing. I'm hoping that others have thoughts on this I tend to take the approach of adding my own timestamps and sorting using array sort. I get individual items with /user/<user_id> and then work with the entire set and filter. Sorry i couldn't be of more help!

**Comment by saurabhabh** [comment]
Thanks for the effort @m-onz . Hopefully some one will give a feed back. I will also try looking at how the 'reverse' is implemented and what parameter it actually uses for reversing .

**Comment by m-onz** [comment]
See this: https://github.com/mafintosh/hyperdb/issues/143 ... seems like if history stream would allow a prefix it would be the ideal solution.

**Comment by saurabhabh** [comment]
Thanks. Will give it a try.

**Comment by saurabhabh** [comment]
Just to update @m-onz  I am on the latest version of lib i.e. 3.5.0 and the history stream does not work with prefixing. If we give some prefix, it still returns **all the nodes** historically sorted.

**Comment by m-onz** [comment]
I misread the issue that I linked too.... 

from @e-e-e 

> In the latest release you can now use createKeyHistoryStream(key) to get the history of a specific key. If you want to get the history a prefix recursively - this is not to difficult to achieve by combining get keyHistory and createReadStream.

You can use ```createKeyHistoryStream``` in the latest version to get the keys your interested in. This has been a bit of an adventure! I can make a code example for this if you need me too.

Hopefully this info is useful to others!

**Comment by saurabhabh** [comment]
Hello, I have implemented it for the time being as follows : 
I store users with timestamp as keys
users/{timestamp1}
users/{timestamp2}
....

I store the time stamps keys in an array 
users/keys/
{
  timestamp1 ,
  timestamp2
}

I sort the keys by timestamp and then get the required users. 
Let me know if you have a better approach.
